integrations/kube-agent-updater/cmd/teleport-kube-agent-updater/main.go

```diff
@@ -166,6 +169,34 @@ func main() {
 		os.Exit(1)
 	}
 
+	// The updater is only designed to work with a single deployment or statefulset.
+	// Pre-fetch the ID if it already exists, so that we can use it to lookup versions.
+	// This logic could be moved into the reconciler, but it would require refactoring
+	// all version getter implementations to accept an ID dynamically in GetVersion.
+	var updateID uuid.UUID
+	var updateYAML v1.ConfigMap
+	err = mgr.GetAPIReader().Get(ctx, kclient.ObjectKey{
+		Namespace: agentNamespace,
+		Name:      agentName + "-updater",
+	}, &updateYAML)
+	switch {
+	case apierrors.IsNotFound(err):
+		updateID, err = uuid.NewRandom()
+		if err != nil {
+			ctrl.Log.Error(err, "failed to generate updater ID, exiting")
+			os.Exit(1)
+		}
+		ctrl.Log.Info("generated new updater ID", "id", updateID)
+	case err != nil:
+		ctrl.Log.Error(err, "failed to generate updater ID, exiting")
+		os.Exit(1)
+	default:
+		updateID, err = uuid.Parse(updateYAML.Data[agent.BinaryName+".id"])
+		if err != nil {
+			ctrl.Log.Error(err, "updater configmap is malformed, canary deployment may fail", "configmap", updateYAML)
+		}
+	}
+
```

Comment:
I think we must create the configmap with the updater ID as soon as we have generated it, instead of waiting for the first update cycle.

Else we might end up with 2 updaters generating different UUIDs and racing for reconciliation, causing unpredictable reported UUID.

This way, the updater can die in case of failed create and re-read the UUID.

Reply:
> Else we might end up with 2 updaters generating different UUIDs and racing for reconciliation, causing unpredictable reported UUID.

Not sure I understand. Two controllers should never reconcile the same statefulset or deployment, as far as I understand.

Reply:
It looks like leader election is enabled for this controller unless you explicit disable it with a flag (which has an outside-of-kubernetes-only warning). The leader election ID is set to value that should be the same during updates.

Multiple instances of the same controller reconciling the same resource is usually considered an undefined behavior state. It's very difficult to write controllers that account for this condition.

Reply:
Addressed in 0f4c432bf822712fc2825dba1b86345957500411